pilot/pkg/model/gateway.go

```diff
@@ -0,0 +1,61 @@
+// Copyright 2017 Istio Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package model
+
+import (
+	"reflect"
+	"strings"
+
+	routing "istio.io/api/routing/v1alpha2"
+)
+
+// MergeGateways takes two Gateways and combines their Servers. When two servers are the same (serversEqual == true)
+// the hosts exposed by both servers are merged. Otherwise, the server is added to dst's server set.
```

Comment:
Q: how will this work in routing rules, where I can refer to ports by name? the merged server has only one port name. 

I suggest you qualify this comment with the specific use case scenario (like ingress and istio gateway, etc.).

Reply:
Or may be say that the ports must have same names as well.

Reply:
Meh, I forgot we supported both names and numbers. I thought we required numbers and allowed names. With the PR in its current state we collect Gateway names outside of this method; we could do the same for port names too, but it gets a bit unwieldy.